src/lib.rs

```diff
@@ -0,0 +1,108 @@
+
+extern crate time;
+extern crate url;
+
+use std::collections::HashMap;
+use time::Timespec;
+use url::Url;
+
+#[derive(Clone, Debug)]
+pub struct PlaceAction {
+    pub url: Url,
+    pub when: Timespec,
+    pub event: String,
+    // TODO: Think about more complex types -- Or do we want callers just
+    // storing JSON here? Can we use std::any::Any and hope to serialize it?
+    pub data: String
+}
+
+impl PlaceAction {
+    pub fn new(url: Url, event: &str, data: &str) -> PlaceAction {
+        PlaceAction {
+            url: url,
+            when: time::get_time(),
+            event: String::from(event),
+            data: String::from(data),
+        }
+    }
+}
+
+pub trait Store {
+    fn record_place_action(&mut self, action: PlaceAction);
+    fn query_place_actions<'a>(&'a self, url: &Url) -> &'a [PlaceAction];
+}
+
+#[derive(Clone, Debug)]
+pub struct MemoryStore {
+    data: HashMap<Url, Vec<PlaceAction>>,
+    // This allows us to avoid needing to put Option<&'a [PlaceAction]>
+    // as the return type for query_place_actions (where None would be used for
+    // "no actions"). Instead, we just return a reference to a private sentinel
+    // vector, which we ensure is always empty.
+    sentinel: Vec<PlaceAction>,
+}
+
+impl MemoryStore {
+    pub fn new() -> MemoryStore {
+        MemoryStore {
+            data: HashMap::new(),
+            sentinel: Vec::new(),
+        }
+    }
+}
+
+impl Store for MemoryStore {
+    fn record_place_action(&mut self, action: PlaceAction) {
+        if let Some(v) = self.data.get_mut(&action.url) {
+            v.push(action);
+            return;
+        }
+        // Can't use else since it would be 2 simultaneous mutable borrows
+        self.data.insert(action.url.clone(), vec![action]);
+    }
+
+    fn query_place_actions<'a>(&'a self, url: &Url) -> &'a [PlaceAction] {
```

Comment:
I wonder if we should return an owned copy of the `Vec`, instead of tying it to the store's lifetime. On the one hand, that's inefficient; on the other, I wonder how we're going to enforce this in the C++/Swift wrapper. Maybe we can have another layer do the cloning...or maybe this will all work fine as is!

Reply:
My concern is that it could be an enormous number of records, since we do no filtering. Consider all the actions users might have taken on websites like facebook or reddit.

Given that, I think even returning a slice is bad here. We probably should be returning an iterator of some sort (or a cursor)? But I don't know how we'd do that with FFI either.

Reply:
Honestly, I wouldn't worry about slices and iterators and cloning and ownership at this stage.

If you're just exploring concepts, you don't need efficiency. If you're not just exploring API modeling concepts, and you're worried about scale, then you don't want a string-based key-value store anyway.

I'd be surprised if a consumer ever had a good reason to call `query_place_actions`.

You might be interested in how SQLite on Android uses cursor windows.